kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,1338 @@
+        ReorderIndex {
+            index,
+            transform: ReorderIndexTransform::None,
+        }
+    }
+
+    fn new_missing(index: usize, field: ArrowFieldRef) -> Self {
+        ReorderIndex {
+            index,
+            transform: ReorderIndexTransform::Missing(field),
+        }
+    }
+
+    /// Check if this reordering requires a transformation anywhere. See comment below on
+    /// [`is_ordered`] to understand why this is needed.
+    fn needs_transform(&self) -> bool {
+        match self.transform {
+            // if we're casting or inserting null, we need to transform
+            ReorderIndexTransform::Cast(_) | ReorderIndexTransform::Missing(_) => true,
+            // if our children are not ordered somehow, we need a transform
+            ReorderIndexTransform::Child(ref children) => !is_ordered(children),
+            // no transform needed
+            ReorderIndexTransform::None => false,
+        }
+    }
+}
+
+// count the number of physical columns, including nested ones in an `ArrowField`
+fn count_cols(field: &ArrowField) -> usize {
+    _count_cols(field.data_type())
+}
+
+fn _count_cols(dt: &ArrowDataType) -> usize {
+    match dt {
+        ArrowDataType::Struct(fields) => fields.iter().map(|f| count_cols(f)).sum(),
+        ArrowDataType::Union(fields, _) => fields.iter().map(|(_, f)| count_cols(f)).sum(),
+        ArrowDataType::List(field)
+        | ArrowDataType::LargeList(field)
+        | ArrowDataType::FixedSizeList(field, _)
+        | ArrowDataType::Map(field, _) => count_cols(field),
+        ArrowDataType::Dictionary(_, value_field) => _count_cols(value_field.as_ref()),
+        _ => 1, // other types are "real" fields, so count
+    }
+}
+
+/// helper function, does the same as `get_requested_indices` but at an offset. used to recurse into
+/// structs, lists, and maps. `parquet_offset` is how many parquet fields exist before processing
+/// this potentially nested schema. returns the number of parquet fields in `fields` (regardless of
+/// if they are selected or not) and reordering information for the requested fields.
+fn get_indices(
+    start_parquet_offset: usize,
+    requested_schema: &Schema,
+    fields: &Fields,
+    mask_indices: &mut Vec<usize>,
+) -> DeltaResult<(usize, Vec<ReorderIndex>)> {
+    let mut found_fields = HashSet::with_capacity(requested_schema.fields.len());
+    let mut reorder_indices = Vec::with_capacity(requested_schema.fields.len());
+    let mut parquet_offset = start_parquet_offset;
+    for (parquet_index, field) in fields.iter().enumerate() {
+        debug!(
+            "Getting indices for field {} with offset {parquet_offset}, with index {parquet_index}",
+            field.name()
+        );
+        match field.data_type() {
+            ArrowDataType::Struct(fields) => {
+                if let Some((index, _, requested_field)) =
+                    requested_schema.fields.get_full(field.name())
+                {
```

Comment:
If I'm not mistaken, we currently have significant code replication, because every match arm includes the same requested field extraction:
```rust
                if let Some((index, _, requested_field)) =
                    requested_schema.fields.get_full(field.name())
                {
                    ...
                }
```
seems better to do the extraction first, and then match only if successful?
```rust
        if let Some((index, _, requested_field)) =
            requested_schema.fields.get_full(field.name())
        {
            match field.data_type() {
                ArrowDataType::Struct(fields) => { ... }
                  ...
                _ => { ... }
            }
        }
```
could even take it a step further with a `filter_map` call:
```rust
    let requested_fields = fields
        .iter()
        .enumerate()
        .filter_map(|parquet_index, field| {
            debug!(
                "Getting indices for field {} with offset {parquet_offset}, with index {parquet_index}",
                field.name(),
            )
            requested_schema.fields.get_full(field.name())
        });
    for (index, _, requested_field) in requested_fields {
```

Reply:
Yeah, I was able to improve this thanks to this suggestion. Couldn't quiet do what you suggested because even for unselected fields we need info (so we can skip their leaf columns), but now I can make an iterator that gets out the needed info, and then iterate over that and reduce the match complexity quite a bit.